### DISCUSSION TITLE
Has anyone ever thought about a plugin to integrate Kimai with FastBill?

### DISCUSSION THREAD

**Discussion by heiglandreas** [kimai/kimai#3793] [Plugins]
I'm using FastBill for my invoicing etc. And while FastBill includes a way to track times I find it rather unintuitive especially when working with multiple clients. So my idea was to use an integration with FastBill that uses their API to populate the clients and projects and then at the end of the month or on the push of a button all of Kimais entries would be copied over to FastBill via the API again so that I am able to use FastBills invoicing features. 

Am I the first one thinking about that? 

I didn't see a plugin in the marketplace (or did I just not look hard enough?) or found issues or discussions regarding that.

So I thought before starting that project myself, I'd rather ask...

**Comment by kevinpapst** [comment]
There are not many integrations yet with third-party products. All I know of are internal ones, tailored to the customer needs. 

Here are some feature requests for a more general approach:
https://github.com/kimai/kimai/issues?q=is%3Aopen+is%3Aissue+label%3A%223-party+integration%22

I'd love to start a new plugin, which adds these missing connections. Shall we do it together?

**Reply by heiglandreas** [reply]
Sounds like a plan :wink:

Though I do see a challenge in the "general" approach as each integration needs their own work to be done. So we can of course provide one common interface (which would be "the API") but then for each third-party one would need a way to convert "the API" into the respective third-party requirement.

Unless all the third-party providers share one common way of interacting with them.

I see mainly two different directions and necessities:

1. Sync the third-partys information to populate the customers, projects and activities
2. Sync kimais timetracking information back to the third-party.

So while that is currently already possible using for example Zapier, that still requires some manual work to do the integration. So having a plugin that does that out-of-the-box would indeed be great!

And I see two main topics that would need to be handled for each third party. One would be to handle the respective API endpoints and one would be to handle the different ways to handle the request and response-formating.

**Reply by kevinpapst** [reply]
Ah, yeah, written language is full of possible misunderstandings. without sufficient explanation 😁 With "general" I meant that we should write an implementation, that can be used by everyone providing the correct configurations.

For example: I have a Tiger CRM sync here, which makes several assumptions for the specific customer, who originally paid for the implementation. This needs to be made configurable to be "general usable" and only then I can add it to that not-yet-existing 3rd `ProviderBundle` (ideas for the bundle name are welcome, SyncBundle?)

Might be best to start with concrete implementations for one provider, then add another one and then check if we see possible code abstractions/simplifications. 

And I would start with your point 2. Way more people asked for that (including yourself).

Kimai bundles are described in the documentation, but they are basically a Symfony bundle with some stuff around. Should be fairly easy to start by copy&pasting the https://github.com/Keleo/DemoBundle as I haven't updated the docs for the upcoming next major version yet 🤷 But the main branch of DemoBundle is compatible with this repos main branch.

**Reply by heiglandreas** [reply]
IMO we need the first part as well as there is no real way to sync the timesheet entries back without the information about customers, projects and activities. Not all third party integrations might need all of those but some do. 

And for me that would be essential :grin:

**Reply by kevinpapst** [reply]
If that is a requirement in the API then yes and it seems in the FastBill API it is. 
Then a full sync is probably the simplest solution first.
